### PR TITLE
new features

### PR DESCRIPTION
**Pull request by aikramer2**
- alternative methods for calculating feature importance
- improvements to datamanager
- optional training labels
- feature importance sampling

### CODE REVIEW

**Comment by aikramer2** [comment]
https://github.com/datascienceinc/Skater/issues/182
https://github.com/datascienceinc/Skater/issues/159
https://github.com/datascienceinc/Skater/issues/152

**Comment by aikramer2** [comment]
also solves https://github.com/datascienceinc/Skater/issues/151

**Comment by aikramer2** [comment]
given the number of changes this will bump the version from 1.0.2 to 1.0.3

**Comment by aikramer2** [comment]
need to add new unit tests for the new methods provided

File: skater/core/explanations.py [hunk 1]
```diff
@@ -53,6 +53,10 @@ def __init__(self, training_data=None, feature_names=None, index=None,
         self.partial_dependence = PartialDependence(self)
         self.feature_importance = FeatureImportance(self)
         self.data_set = None
+        self.training_labels = None
+        if training_labels is not None:
+            self.training_labels = DataManager(training_labels, index=index)
+
```
Comment (pramitchoudhary) [inline skater/core/explanations.py]: Avoidable blank lines

File: skater/core/global_interpretation/feature_importance.py [hunk 2]
```diff
@@ -3,71 +3,25 @@
 import numpy as np
 import pandas as pd
 from functools import partial
-from pathos.multiprocessing import Pool
```
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: Why are we getting rid of multiprocessing provided by pathos ? or is it just a rename of the pathos pkg ?

File: skater/core/global_interpretation/feature_importance.py [hunk 3]
```diff
@@ -233,40 +246,136 @@ def plot_feature_importance(self, predict_fn, filter_classes=None, ascending=Tru
...
-        # to resolve the issuse of sorting based on axis
-        if ascending is True:
-            importances.sort_values(ascending=False).plot(kind='barh', ax=ax, color=color)
-        else:
-            importances.sort_values(ascending=True).plot(kind='barh', ax=ax, color=color)
+        importances.sort_values(ascending=ascending).plot(kind='barh', ax=ax, color=color)
         return f, ax
 
+    @staticmethod
+    def compute_feature_importance(feature_id, input_data, estimator_fn,
```
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: We might have to do more tests to validate that when one increases the number of jobs, the computation succeeds. I remember we tried parallelizing using staticmethod in the past, and it failing in-consistently. 
* To start, lets also add a test with 2 jobs and validate that things succeed 
* To be 100% sure, lets also check, one of the existing notebooks

File: skater/core/global_interpretation/feature_importance.py [hunk 4]
```diff
@@ -115,41 +85,61 @@ def feature_importance(self, model_instance, ascending=True, filter_classes=None
             filter_classes = list(filter_classes)
             assert all([i in model_instance.target_names for i in filter_classes]), err_msg
 
-        original_predictions = model_instance.predict(self.data_set.data)
+        if method == 'performance-decrease' and self.training_labels is None:
+            raise FeatureImportanceError("If interpretation.training_labels are not set, you"
+                                         "can only use feature importance methods that do "
+                                         "not require ground truth labels")
+        elif method == 'performance-decrease':
```
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: We might be able to do a better job here, if the target_label = None, just checking the existing of that, might help us know what method to use, and one can flag error accordingly

File: skater/core/global_interpretation/feature_importance.py [hunk 5]
```diff
@@ -115,41 +85,61 @@ def feature_importance(self, model_instance, ascending=True, filter_classes=None
...
+                                                   sample=True,
+                                                   n_samples=n_samples)
+        else:
+            inputs = self.data_set.data
 
-        n = original_predictions.shape[0]
+        original_predictions = model_instance.predict(inputs)
+        model_type = model_instance.model_type
 
         if progressbar:
```
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: the check on whether progressbar is enabled seems to be all over the code base. 
Lets see, if it is possible to do better there. May be abstract is out as a function ?

File: skater/core/explanations.py (hunk 1 above)
Comment (aikramer2) [inline skater/core/explanations.py]: fixed

File: skater/core/global_interpretation/feature_importance.py (hunk 2 above)
Comment (aikramer2) [inline skater/core/global_interpretation/feature_importance.py]: i did some benchmarking, and got consistently better performance out of multiprocess

File: skater/core/global_interpretation/feature_importance.py [hunk 6]
```diff
@@ -87,6 +42,21 @@ def feature_importance(self, model_instance, ascending=True, filter_classes=None
...
             Only used in classification models.
+        n_jobs: int
+            How many concurrent processes to use. Defaults -1, which grabs as many as are available.
+            Use 1 to avoid multiprocessing altogether.
+        progressbar: bool
+            Whether to display progress. This affects which function we use to operate on the pool
+            of processes, where including the progress bar results in 10-20% slowdowns.
+        n_samples: int
+            How many samples to use when computing importance.
+        method: string
```
Comment (aikramer2) [inline skater/core/global_interpretation/feature_importance.py]: ya its poorly named. The Wei 2015 paper refers to this class of methods as condition-permutation methods, so lets go with that.

File: skater/core/global_interpretation/feature_importance.py (hunk 4 above)
Comment (aikramer2) [inline skater/core/global_interpretation/feature_importance.py]: id like to avoid that, to ensure the user can have control.

File: skater/core/global_interpretation/feature_importance.py (hunk 5 above)
Comment (aikramer2) [inline skater/core/global_interpretation/feature_importance.py]: i revised a little to minimize the number of checks of the boolean `progressbar`.

File: skater/core/global_interpretation/feature_importance.py (hunk 3 above)
Comment (aikramer2) [inline skater/core/global_interpretation/feature_importance.py]: its now part of the test suite

File: skater/model/scorer.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        Parameters
+        ------------
+        predictions: array type
+        labels: array type
+        model_instance: subtype of skater.util.static_types.ModelTypes
+        """
+        assert predictions.shape[0] == labels.shape[0], "labels and prediction shapes dont match" \
+                                                        "predictions: {0}, labels: {1}".format(predictions.shape,
+                                                                                               labels.shape)
+        # continuous, binary, continuous multioutput, multiclass, multilabel-indicator
```
Comment (aikramer2) [inline skater/model/scorer.py]: added link to the function's documentation, which includes all the explanations

File: skater/model/scorer.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
+from sklearn.metrics import log_loss, mean_absolute_error
+from sklearn.utils.multiclass import type_of_target
+
+from ..util.static_types import StaticTypes
+
+
+class Scorer(object):
```
Comment (aikramer2) [inline skater/model/scorer.py]: I think Im going to redesign the scorer a bit to accomplish a few things:
1) The scorer is tied to the model. Each model gets a scorer object, which can be used throughout the lifecycle of the model.
2) User defines the metric, the scorer does everything else (checking for compatibility, picking implementation, etc.)
3) scorer, like model objects, provide a `build_static_scorer` function, which only takes inputs, labels, and metric arguments, so the the scorer can be executed in a multi-processing setting.

File: skater/core/explanations.py [hunk 9]
```diff
@@ -35,7 +35,7 @@ class Interpretation(object):
     interpreter.partial_dependence([feature_id1, feature_id2], regressor.predict)
     """
 
-    def __init__(self, training_data=None, feature_names=None, index=None,
+    def __init__(self, training_data=None, training_labels=None, feature_names=None, index=None,
```
Comment (aikramer2) [inline skater/core/explanations.py]: i want to keep as training_labels for now, since theres an obvious relationship between training_labels and training_data, in that they be corresponding rows. Its also clear that the labels should be those that were used during model training.

**Comment by aikramer2** [comment]
thanks for taking a look @pramitchoudhary. Some of the smaller changes are done on my local machine. The only other major thing for me to do before pushing to this PR is to design the scorer according to my comments above^^.

**Comment by pramitchoudhary** [comment]
thanks @aikramer2. Looking into it, will add more comments and suggestion as I further look into. 
This would be an awesome story :)

File: skater/core/explanations.py (hunk 9 above)
Comment (pramitchoudhary) [inline skater/core/explanations.py]: May be, but if one has target_labels related to test data is there anything that will stops us from evaluating the model on that. And this will flow nicely with the deploy story. 
Think about it

File: skater/core/global_interpretation/feature_importance.py (hunk 2 above)
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: ya, but pathos might have a better serialization structure. 
Also, would recommend to do changes iteratively mean this change has too many things going,
* Re-structuring
* swapping out libraries
* new feature
A lot of high risk changes. Only if we have tested throughly can we allow the switch of library.
If its not affecting us in anyway. Lets get feature and re-structuring in, and then swap out the library, might help in minimizing the risk.

File: skater/core/global_interpretation/feature_importance.py (hunk 6 above)
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: possible to provide the link to that paper ?

File: skater/core/global_interpretation/feature_importance.py (hunk 4 above)
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: Like i had mentioned earlier, we might not need to have a flag explicitly for this, based on the existence of the `target_label`!=None one can decide which method to use. This might also avoid un-necessary variable combinations

File: skater/core/global_interpretation/feature_importance.py (hunk 5 above)
Comment (pramitchoudhary) [inline skater/core/global_interpretation/feature_importance.py]: cool, 👍 will check

File: skater/model/scorer.py (hunk 8 above)
Comment (pramitchoudhary) [inline skater/model/scorer.py]: I like the 3 point, not sure on the first 2. I feel that if we structure that as dictionary construct, it might help a lot. Possible to first try with simple implementation before adding abstraction ?
Actually, before re-structuring the code, lets do a quick design review. Something simple, just a basic design on a paper articulating the thought process will be helpful and will avoid re-work.

File: skater/tests/test_feature_importance.py [hunk 10]
```diff
@@ -72,9 +72,62 @@ def feature_column_name_formatter(columnname):
...
     def test_feature_importance(self):
-        importances = self.interpreter.feature_importance.feature_importance(self.regressor_predict_fn)
-        self.assertEquals(np.isclose(importances.sum(), 1), True)  # default grid resolution is 100
-
+        importances = self.interpreter.feature_importance.feature_importance(self.regressor_predict_fn,
+                                                                             progressbar=False)
+        self.assertEquals(np.isclose(importances.sum(), 1), True)
+
+
+    def test_feature_importance_progressbar(self):
```
Comment (pramitchoudhary) [inline skater/tests/test_feature_importance.py]: Like the test coverage here, nice work @aikramer2 
Checking on implementation

File: skater/tests/test_feature_importance.py [hunk 11]
```diff
@@ -72,9 +72,62 @@ def feature_column_name_formatter(columnname):
```
Comment (pramitchoudhary) [inline skater/tests/test_feature_importance.py]: Possible to add an extra blank line between to unit functions. We are following a similar pattern for our regular functions
